MonitorControl/Model/AppleDisplay.swift

```diff
@@ -0,0 +1,72 @@
+//  Copyright © MonitorControl. @JoniVR, @theOneyouseek, @waydabber and others
+
+import Foundation
+import os.log
+
+class AppleDisplay: Display {
+  private var displayQueue: DispatchQueue
+
+  override init(_ identifier: CGDirectDisplayID, name: String, vendorNumber: UInt32?, modelNumber: UInt32?, isVirtual: Bool = false) {
+    self.displayQueue = DispatchQueue(label: String("displayQueue-\(identifier)"))
+    super.init(identifier, name: name, vendorNumber: vendorNumber, modelNumber: modelNumber, isVirtual: isVirtual)
+  }
+
+  public func getAppleBrightness() -> Float {
+    var brightness: Float = 0
+    DisplayServicesGetBrightness(self.identifier, &brightness)
+    return brightness
+  }
+
+  public func setAppleBrightness(value: Float) {
+    self.displayQueue.sync {
+      DisplayServicesSetBrightness(self.identifier, value)
+      DisplayServicesBrightnessChanged(self.identifier, Double(value))
+    }
+  }
+
+  override func setDirectBrightness(_ to: Float, transient: Bool = false) -> Bool {
+    let value = max(min(to, 1), 0)
+    self.setAppleBrightness(value: value)
+    if !transient {
+      self.savePref(value, for: .brightness)
+      self.brightnessSyncSourceValue = value
+      self.smoothBrightnessTransient = value
+    }
+    return true
+  }
+
+  override func getBrightness() -> Float {
+    if self.prefExists(for: .brightness) {
+      return self.readPrefAsFloat(for: .brightness)
+    } else {
+      return self.getAppleBrightness()
+    }
+  }
+
+  override func refreshBrightness() -> Float {
+    guard !self.smoothBrightnessRunning else {
+      return 0
+    }
+    let brightness = self.getAppleBrightness()
+    let oldValue = self.brightnessSyncSourceValue
+    self.savePref(brightness, for: .brightness)
+    if brightness != oldValue {
+      os_log("Pushing slider and reporting delta for Apple display %{public}@", type: .debug, String(self.identifier))
+      var newValue: Float
+
+      if abs(brightness - oldValue) < 0.01 {
+        newValue = brightness
+      } else if brightness > oldValue {
+        newValue = oldValue + max((brightness - oldValue) / 3, 0.005)
+      } else {
+        newValue = oldValue + min((brightness - oldValue) / 3, -0.005)
```

Comment:
it is half of 1% which is the minimum possible difference the user can set using keyboard if he uses fine control. I set this as a minimum speed (minimum level of change in each iteration) as the smoothing logic approaches the desired target brightness to avoid situations when it goes into standstill or creeps forever (but this is also prevented by the fact that once the brightness reaches a lower than 1% difference to the target brightness, line 57-58 kicks in and the process is over).